### PR TITLE
ENH: Support qualified names in update_page_form_field_values

### PR DESCRIPTION
**Pull request by xi**
PDF forms often use names like "A.1", "A.2", "B.1", "B.2", … for the fields. However, the `.` has a special meaning, so this creates a hierarchy instead.

It was impossible to fill those individual fields with `update_page_form_field_values()`:

```python
update_page_form_field_values({"A": "foo"})  # fills all "A.*" fields
update_page_form_field_values({"1": "foo"})  # fills all "*.1" fields
update_page_form_field_values({"A.1": "foo"})  # fills none of the fields
```

This change makes `update_page_form_field_values()` to also check for qualified field names.

See also #545

---

@hchillon took a different approach in #545. Unfortunately I don't know enough about PDF to fully understand what they did there. This simple approach worked well for my usecase.

I copied `_get_qualified_field_name()` from the `PdfReader` class. Should this be moved to a utility module so it can be shared in both classes? Where would that be?

### CODE REVIEW

**Comment by pubpub-zz** [comment]
As you have noticed, period is not accepted in field names
(extract for pdf standard 1.7 App H, page 1117)
image

Your idea to extend(modify?) to take into account qualified.names sounds good  and I like your idea to use `_get_qualified_filed_name()` 👍. I don't think we need an extra param to force to check standard names only

use the same name _get_qualified_filed_name in writer (later we may be able to merge some code)

**Comment by pubpub-zz** [comment]
@xi,
can you fix the mypy issues ?

File: pypdf/_writer.py [hunk 1]
```diff
@@ -795,11 +812,14 @@ def update_page_form_field_values(
         for j in range(len(page[PG.ANNOTS])):  # type: ignore
             writer_annot = page[PG.ANNOTS][j].get_object()  # type: ignore
             # retrieve parent field values, if present
-            writer_parent_annot = {}  # fallback if it's not there
+            writer_parent_annot = cast(DictionaryObject, {})  # fallback if it's not there
```
Comment (MartinThoma) [inline pypdf/_writer.py:815]: Is there a disadvantage in just creating that object?

```suggestion
            writer_parent_annot = DictionaryObject()  # fallback if it's not there
```

**Comment by MartinThoma** [comment]
Thank you for your contribution @xi :pray: 

To be honest, I'm also always struggling with the form-stuff. It might take a while until I decided on how to continue with this PR. Any input (e.g. @pubpub-zz / @MasterOdin / others who know more about PDF) is very welcome :-)

**Comment by pubpub-zz** [comment]
I added a comment about earlier and for me this is good.

**Comment by pubpub-zz** [comment]
@xi 
Just another point : can you add tests for test coverage

**Comment by xi** [comment]
>  Just another point : can you add tests for test coverage

I don't think I know enough to create a minimal PDF that demonstrates this issue. Can someone please help out?

**Comment by pubpub-zz** [comment]
@xi 
I've started to copy/upgrade test_fill_form() in test_writer.py:
```
def test_fill_form_with_qualified():
    reader = PdfReader(RESOURCE_ROOT / "form.pdf")
    reader.r.add_form_topname("top")
    writer = PdfWriter()

    page = reader.pages[0]

    writer.add_page(page)
    writer.update_page_form_field_values(
        writer.pages[0], {"top.foo": "filling"}, flags=1
    )

    b = BytesIO()
    writer.write(0)

    reader2.get_fields["top.foo"]  # to be completed....
```

My dev environment is not available for the moment to finalize the code. Can you do it please

**Comment by xi** [comment]
@pubpub-zz thanks for the head start.  I tried to fill in the gaps in https://github.com/py-pdf/pypdf/pull/1695/commits/bc0458dbe7ee2180e3cbd6daf11c423e64070648.

**Comment by MartinThoma** [comment]
Thank you for your contribution @xi :pray:  It will be in `pypdf > 3.5.2`, which I will likely release on Sunday.

If you want, I can add you to the contributors: https://pypdf.readthedocs.io/en/latest/meta/CONTRIBUTORS.html

Just let me know which name I should use / what I should link to :-)

**Pull request merged**